repo/fsrepo/datastores.go

```diff
@@ -0,0 +1,330 @@
+)
+
+// ConfigFromMap creates a new datastore config from a map
+type ConfigFromMap func(map[string]interface{}) (DatastoreConfig, error)
+
+// DatastoreConfig is an abstraction of a datastore config.  A "spec"
+// is first converted to a DatastoreConfig and then Create() is called
+// to instantiate a new datastore
+type DatastoreConfig interface {
+	// DiskSpec returns a minimal configuration of the datastore
+	// represting what is stored on disk.  Run time values are
+	// excluded.
+	DiskSpec() DiskSpec
+
+	// Create instantiate a new datastore from this config
+	Create(path string) (repo.Datastore, error)
+}
+
+// DiskSpec is the type returned by the DatastoreConfig's DiskSpec method
+type DiskSpec map[string]interface{}
+
+// Bytes returns a minimal JSON encoding of the DiskSpec
+func (spec DiskSpec) Bytes() []byte {
+	b, err := json.Marshal(spec)
+	if err != nil {
+		// should not happen
+		panic(err)
+	}
+	return bytes.TrimSpace(b)
+}
+
+// String returns a minimal JSON encoding of the DiskSpec
+func (spec DiskSpec) String() string {
+	return string(spec.Bytes())
+}
+
+var datastores map[string]ConfigFromMap
+
+func init() {
+	datastores = map[string]ConfigFromMap{
+		"mount":   MountDatastoreConfig,
+		"flatfs":  FlatfsDatastoreConfig,
+		"levelds": LeveldsDatastoreConfig,
+		"mem":     MemDatastoreConfig,
+		"log":     LogDatastoreConfig,
+		"measure": MeasureDatastoreConfig,
+	}
+}
+
+// AnyDatastoreConfig returns a DatastoreConfig from a spec based on
+// the "type" parameter
+func AnyDatastoreConfig(params map[string]interface{}) (DatastoreConfig, error) {
+	which, ok := params["type"].(string)
+	if !ok {
+		return nil, fmt.Errorf("'type' field missing or not a string")
+	}
+	fun, ok := datastores[which]
+	if !ok {
+		return nil, fmt.Errorf("unknown datastore type: %s", which)
+	}
+	return fun(params)
+}
+
+type mountDatastoreConfig struct {
+	mounts []premount
+}
+
+type premount struct {
+	ds     DatastoreConfig
+	prefix ds.Key
+}
+
+// MountDatastoreConfig returns a mount DatastoreConfig from a spec
+func MountDatastoreConfig(params map[string]interface{}) (DatastoreConfig, error) {
+	var res mountDatastoreConfig
+	mounts, ok := params["mounts"].([]interface{})
+	if !ok {
+		return nil, fmt.Errorf("'mounts' field is missing or not an array")
+	}
+	for _, iface := range mounts {
+		cfg, ok := iface.(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("expected map for mountpoint")
+		}
+
+		child, err := AnyDatastoreConfig(cfg)
+		if err != nil {
+			return nil, err
+		}
+
+		prefix, found := cfg["mountpoint"]
+		if !found {
+			return nil, fmt.Errorf("no 'mountpoint' on mount")
+		}
+
+		res.mounts = append(res.mounts, premount{
+			ds:     child,
+			prefix: ds.NewKey(prefix.(string)),
+		})
+	}
+
+	return &res, nil
+}
+
+func (c *mountDatastoreConfig) DiskSpec() DiskSpec {
+	cfg := map[string]interface{}{"type": "mount"}
+	mounts := make([]interface{}, len(c.mounts))
```

Comment:
Noticed one more small thing: mounts in spec should be sorted as the order doesn't matter, see https://github.com/ipfs/go-datastore/blob/master/syncmount/mount.go#L31